Jenkinsfile

```diff
@@ -92,56 +92,149 @@ pipeline {
+              }
+            }
+            script {
+              if (env.GCP_ENV_ERROR == "true") {
+                error('GCP_ENV_ERROR cannot deploy function')
+              }
 
-                stash name: 'authnGcpTokens', includes: 'gcp_token_valid,gcp_token_invalid_audience,gcp_token_standard_format,gcp_token_user,gcp_token_non_existing_host,gcp_token_non_existing_account,gcp_token_non_rooted_host', allowEmpty:false
-                env.GCP_TOKENS_FETCHED = "true"
+              dir('ci/authn-gcp') {
+                sh '''
+                # Deploy Google Cloud function
+                summon ./run_gcloud.sh deploy_function.sh || exit 1
+                if [ $? -ne 0 ]; then
+                    echo '-- Error deploying Google function'
+                    exit 1
+                fi
+                # Obtain tokens from Google function and write to files
+                ./get_func_tokens_to_files.sh
+                if [ $? -ne 0 ]; then
+                    echo '-- Error obtaining tokens from Google function'
+                    exit 1
+                fi
+                '''
+              }
+            }
+          }
+          post {
+            success {
+              echo "-- Google Cloud test env is ready"
+              script {
+                env.GCP_FUNC_TOKENS_FETCHED = "true"
+              }
+            }
+            failure {
+              echo "-- GCP function deployment stage failed"
+              script {
+                env.GCP_ENV_ERROR = "true"
+              }
+            }
+            always {
+              script {
+                dir('ci/authn-gcp') {
+                  sh '''
+                  # Cleanup Google function
+                  summon ./run_gcloud.sh cleanup_function.sh
+                  '''
+                }
               }
             }
           }
         }
-        stage('GCP Authenticator') {
+        /**
+        * We have 3 stages for GCP Authenticator tests.
+        * This stage waits for GCP Stage A and B to complete, unstashes the tokens created in
+        * stage: 'GCP - Stage A - Compute Engine' and runs the gcp-authn tests.
+        */
+        stage('GCP Authenticator - Run tests') {
           steps {
+            echo 'Waiting for GCP_TOKENS'
             timeout(time: 10, unit: 'MINUTES') {
               waitUntil {
                 script {
-                  return (env.GCP_TOKENS_FETCHED == "true")
+                  return ( env.GCP_ENV_ERROR == "true"
+                    || (env.GCP_FUNC_TOKENS_FETCHED == "true" && GCE_TOKENS_FETCHED == "true"))
                 }
               }
             }
+            script {
+              if (env.GCP_ENV_ERROR == "true") {
+                error('GCP_ENV_ERROR cannot run tests check the logs for errors in GCP stages A and B')
+              }
+            }
             script {
               dir('ci/authn-gcp/tokens') {
-                unstash 'authnGcpTokens'
+                unstash 'authnGceTokens'
```

Comment:
why gce?

Reply:
To distinguish GCE token form function tokens.